### ISSUE TITLE
ecaudit 2.8.0 with cassandra 3.11.12

### ISSUE THREAD

**Issue by dsx0** [Ericsson/ecaudit#193]
Hi!

I've just tested version 2.8.0 with latest cassandra 3.11 update (3.11.12) and get the following error:

`
ERROR [Native-Transport-Requests-3] 2022-03-04 12:30:11,059 ErrorMessage.java:384 - Unexpected exception during request
java.lang.StringIndexOutOfBoundsException: String index out of range: -1
	at java.lang.AbstractStringBuilder.setCharAt(AbstractStringBuilder.java:407)
	at java.lang.StringBuilder.setCharAt(StringBuilder.java:76)
	at com.ericsson.bss.cassandra.ecaudit.entry.PreparedAuditOperation.preparedWithValues(PreparedAuditOperation.java:102)
	at com.ericsson.bss.cassandra.ecaudit.entry.PreparedAuditOperation.bindValues(PreparedAuditOperation.java:83)
	at com.ericsson.bss.cassandra.ecaudit.entry.PreparedAuditOperation.getOperationString(PreparedAuditOperation.java:65)
	at com.ericsson.bss.cassandra.ecaudit.logger.Slf4jAuditLogger.lambda$getAvailableFieldFunctionMap$4(Slf4jAuditLogger.java:101)
`

I see, there is already a (failing) PR.
I just wanted to give you some details about this error.

Sorry for the inconveniences.

Let me know if you want some help for testing.

**Comment by eperott** [comment]
Hey! Thanks for the report.

Actually, this was fixed a few days ago in this patch: https://github.com/Ericsson/ecaudit/commit/f9b254ae1ab6a2cf333d70492cde153c7e18e918

I've created a new release 2.9.0 with the fix in place. It should be available on Maven Central in a while.

Let me know your results!

**Comment by dsx0** [comment]
Hi!

I could test it with the new version but i still get the same error.
I have to investigate more on this error.

It's **working with cqlsh** but not (totally) working with standard datastax-cassandra-driver.

On the client-side i'm using now the latest version of cassandra-driver artifacts (3.11.1):
- cassandra-driver-core
- cassandra-driver-mapping
- cassandra-driver-extras

Facts (what works):
- datastax-PreparedStatments: **WORKS**
- datastax-MappingManager: the Mapper **WORKS**
- datastax-MappingManager: the Accessor **does NOT WORK**

It looks like something changed in cassandra too which forwards the actual statement to your PreparedAuditOperation-Class (The authentication already succeeded).
It seems that the "preparedStatement"-String which will be given to PreparedAuditOperation.java#L88 is empty in case of using an "Accessor" from MappingManager (datastax-driver).

_Dummy Example of my Accessor:_
@Accessor
public interface CarAccessor {

	@Query("SELECT * FROM cars")
	Result<DbCar> getAll();
...


Whole Exception on cassandra-side:
`
java.lang.StringIndexOutOfBoundsException: String index out of range: -1
	at java.lang.AbstractStringBuilder.setCharAt(AbstractStringBuilder.java:407)
	at java.lang.StringBuilder.setCharAt(StringBuilder.java:76)
	at com.ericsson.bss.cassandra.ecaudit.entry.PreparedAuditOperation.preparedWithValues(PreparedAuditOperation.java:102)
	at com.ericsson.bss.cassandra.ecaudit.entry.PreparedAuditOperation.bindValues(PreparedAuditOperation.java:83)
	at com.ericsson.bss.cassandra.ecaudit.entry.PreparedAuditOperation.getOperationString(PreparedAuditOperation.java:65)
	at com.ericsson.bss.cassandra.ecaudit.logger.Slf4jAuditLogger.lambda$getAvailableFieldFunctionMap$4(Slf4jAuditLogger.java:101)
	at java.util.function.Function.lambda$andThen$1(Function.java:88)
	at com.ericsson.bss.cassandra.ecaudit.common.formatter.LogMessageFormatter.lambda$getArgumentsForEntry$1(LogMessageFormatter.java:117)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:546)
	at java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
	at java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:505)
	at com.ericsson.bss.cassandra.ecaudit.common.formatter.LogMessageFormatter.getArgumentsForEntry(LogMessageFormatter.java:118)
	at com.ericsson.bss.cassandra.ecaudit.logger.Slf4jAuditLogger.log(Slf4jAuditLogger.java:129)
	at com.ericsson.bss.cassandra.ecaudit.facade.DefaultAuditor.lambda$performAudit$0(DefaultAuditor.java:104)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at com.ericsson.bss.cassandra.ecaudit.facade.DefaultAuditor.performAudit(DefaultAuditor.java:104)
	at com.ericsson.bss.cassandra.ecaudit.facade.DefaultAuditor.audit(DefaultAuditor.java:76)
	at com.ericsson.bss.cassandra.ecaudit.AuditAdapter.auditPrepared(AuditAdapter.java:134)
	at com.ericsson.bss.cassandra.ecaudit.handler.AuditQueryHandler.processPreparedWithAudit(AuditQueryHandler.java:161)
	at com.ericsson.bss.cassandra.ecaudit.handler.AuditQueryHandler.processPrepared(AuditQueryHandler.java:138)
	at org.apache.cassandra.transport.messages.ExecuteMessage.execute(ExecuteMessage.java:163)
	at org.apache.cassandra.transport.Message$Dispatcher.processRequest(Message.java:688)
	at org.apache.cassandra.transport.Message$Dispatcher.lambda$channelRead0$0(Message.java:594)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$FutureTask.run(AbstractLocalAwareExecutorService.java:162)
	at org.apache.cassandra.concurrent.SEPWorker.run(SEPWorker.java:113)
	at java.lang.Thread.run(Thread.java:748)
`

I would need some logging/tracing in this case.
Should i try to fork and add some debug-information for this to analyse?
Or do you have any idea what could have changed in cassandra? It worked for a really long time.


Please let me know what you think about it!

**Comment by dsx0** [comment]
Hi!

I found the problem after some analytics but don't have a solution yet.

There are big changes in PreparedStatement handling in cassandra version 3.11.12.

The reason why the "preparedStatement" is an empty String in class PreparedAuditOperation:
- Cassandra is caching AND storing PreparedStatements on server-side.
- **after deleting** $CASSANDRA_HOME/data/system/prepared_statements-xyz folder (stopped cassandra before and restarted after) the PreparedStatements **work again** because they have to be recreated
- **after another restart it DOES NOT work again**: cassandra loads the stored PreparedStatements again and the "**rawCQLStatement**" (in Class ParsedStatement.Prepared) **is an empty String** again.
  -> therefore an empty String is passed again to PreparedAuditOperation

The Problem is: all PreparedStatements in "QueryProcessor.preparedStatements" have an empty String in "**rawCQLStatement**" after restarting cassandra.

I think it's maybe a bug in cassandra 3.11.12?
Should i report this bug/behaviour to cassandra?

Please let me know what you think about it.

**Comment by eperott** [comment]
Hey!

Thanks for all the effort you're putting on this. I've tried to reproduce the issue with the tooling that I have at hand, but no luck so far.

I don't think we can consider this a bug in Cassandra itself (unless this breaks vanilla Cassandra functionality as well). ecAudit is pulling some tricks with a ThreadLocal in AuditQueryHandler to get to the raw statement. But looks like there are situations where this doesn't work. I've been browsing Cassandra code, and it looks like it is reading up the raw statement but the logic is a bit hairy so not sure if it is stored in memory in all cases. Would have to put some more time on this.

Just for me to understand, are you able to reproduce this ONLY if you use Accessor mode AND do a restart (without clearing the prepared-statement cache)? Anything that would allow me to reproduce, instructions or code, would be great. And if you want to contribute a fix that would be awesome!

**Comment by eperott** [comment]
Also, after restart, does Cassandra log any line like:
> Preloaded X prepared statements

Do you observe any warning in the logs saying:
> Prepared statement recreation error, removing statement: ...

**Comment by dsx0** [comment]
> Also, after restart, does Cassandra log any line like:
> 
> > Preloaded X prepared statements
> 
> Do you observe any warning in the logs saying:
> 
> > Prepared statement recreation error, removing statement: ...

I got normal message:

> Preloaded 6 prepared statements

No errors or warnings above.

**Comment by dsx0** [comment]
Hi!

I'm already done with creation of that testcase!

The testcase starts a new cassandra-cluster from scratch with ecaudit activated.
It contains a simple testclass which acts as client making SELECT calls.

Sadly i **COULD** reproduce the same behaviour with latest cassandra 3.11.12 and ecaudit 2.9.0.

Please have a look at: [cassandra-ecaudit](https://github.com/dsx0/cassandra-ecaudit)
I have an explaination in README.md

How to reproduce:
- Follow steps in README.md to start cassandra with docker-image
- Use Java-Test-Client in [cassandra-ecaudit-testclient](https://github.com/dsx0/cassandra-ecaudit/tree/main/cassandra-ecaudit-testclient). --> **everthing works fine**
- Stop and start docker container (also explained in README.md)
- Use Java-Test-Client again. --> **everything works EXPECT Accessor**

Your can even debug cassandra and ecaudit plugin by java-remote-debugging on port 1414.

Please let me know if you can reproduce it.
Thx in advance!

**Comment by emolsson** [comment]
I apologize for hijacking this, that was not my intention. I read the ticket a week ago and forgot to read the latest on it before starting on my patch today as one of our users reported this internally. I can see that you have put a lot of effort into reproducing the issue and digging into the problem and would love to get your feedback on the PR if you have the opportunity.

I'm not sure exactly how the Accessor works but I think it might be so that we only actually throw an exception in case there are no bound values in the request as we expect there to be values and start manipulating the string before the ending ']' which is usually ', ' when there are values. This also caused prepared statements without bound values to be logged as `<statement>]` rather than `<statement>[]` and that behavior was also fixed in the PR. Even if the other statements doesn't throw any exceptions they are probably still broken though.

> I think it's maybe a bug in cassandra 3.11.12?
> Should i report this bug/behaviour to cassandra?

I haven't read through the JIRA tickets for the work that changed this so I don't know if this was intended, but I think you can definitely raise a ticket for it as it basically removes the possibility of using the "load statements during startup" feature whenever a plugin tries to use the rawCQLStatement.

**Comment by dsx0** [comment]
Hi @emolsson!

Thank you for your input and PR!

1. I started to test my old scenario with cassandra 3.11.11 to check if it was still working there.
-> works with 3.11.11 but **not with 3.11.12** anymore.
So maybe we really should open a ticket for cassandra.

2. Thx for the PR!
I added another Docker-Build for local ecaudit-builds: [cassandra-ecaudit-docker-local-ecaudit](https://github.com/dsx0/cassandra-ecaudit/tree/main/cassandra-ecaudit-docker-local-ecaudit-build)
and placed my build (from your **issue-193** branch) into the "ecaudit-builds" folder.
-> and it **WORKS!**

So your PR makes it work again!
Please, could we get it into next release?

**Comment by eperott** [comment]
Thank you both for your contributions! Awesome!

I've release ecAudit version 2.9.1 for Cassandra 3.11 flavor. Should be visible on Maven central in a few moments.

@emolsson your patch doesn't merge clean to master branch for 4.0 flavor. Related sections in Cassandra code has been changed a bit, and in fact the statement cash is refactored. Will try to reproduce issue and test fix using @dsx0 docker images. Unless you beat me to it. :)

**Comment by emolsson** [comment]
Thanks a lot for the quick responses!

@eperott I only glanced over the 4.0 code but I don't think this workaround is necessary there at all. The `#preloadPreparedStatements()` uses parseAndPrepare() which only sets an empty rawCQLStatement in case it's an [internal operation](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/cql3/QueryProcessor.java#L438). In those cases I would assume that we don't log an audit operation either way, but I haven't verified that. Feel free to look into this further, I'm probably going to have limited time for it this week.. I did copy the test case I created in `ITVerifyAudit` and the statement seems to persist properly through that "cache restart" mechanism at least even without the patch.

**Comment by dsx0** [comment]
I'm not familiar with cassandra 4.
I guess the configuration will be pretty different?

The Dockerfile in my Test-Image can be easily adjusted to use cassandra 4, but the cassandra configs has to be changed too?

Let me know if you need help building it with cassandra 4.
